### PR TITLE
Custom model for comparing performance with ivadomed's model 

### PR DESCRIPTION
**Pull request by naga-karthik**
This PR aims to reproduce the performance obtained by using ivadomed's UNet3D model. The details on the latest results using ivadomed training can be found [here](https://github.com/ivadomed/model_seg_sci/issues/11#issuecomment-1087609028). The training code heavily borrows from ms-challenge 2021 and it is implemented in PyTorch Lightning instead. Below is the summary of the major changes:

1. Dataset: Can be found under `duke/temp/muena/sci-zurich_preprocessed`. This is a preprocessed dataset that includes only sagittal T2w images cropped around the lesions using axial T2w images' field-of-view as reference. 
2. Model: It is ModifiedUNet3D, the same as the one used in ivadomed. 

The issue currently is that the custom training is unable to get the similar performance as ivadomed's despite having the same hyperparameters. Therefore, the goal of this PR is to brainstorm and get some potential suggestions on where the issues might be.

### CODE REVIEW

**Comment by jcohenadad** [comment]
@naga-karthik thank you for opening that up, few questions:

> it is implemented in PyTorch Lightning instead

Is it the main rationale for borrowing from that code? I'm trying to understand the motivation for _not_ using ivadomed?

> the custom training is unable to get the similar performance as ivadomed's

is performance of custom lower of higher than ivadomed?

**Comment by naga-karthik** [comment]
> is performance of custom lower of higher than ivadomed?

It is lower than what I'm able to get by using ivadomed. 

> I'm trying to understand the motivation for not using ivadomed?

So, here's the thing - There are some ideas that I wanted to implement (ie auto encoder pretraining, mask-rcnn etc.), which would be difficult to implement in ivadomed. Hence, I am trying to implement these separately first to see they if work (before trying to incorporate them into ivadomed). However, in order to get a baseline model, I am trying to obtain the same performance as ivadomed's so that it will be easier to compare with the newer methods. As it stands, there is one model in ivadomed, and there are other models that are custom implemented - difficult to compare performance all under one graph).

File: modeling/main.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+# optimizations
+parser.add_argument('-p', '--precision', default=32, type=int, help='Precision for training')
+parser.add_argument('-ne', '--num_epochs', default=500, type=int, help='Number of epochs for the training process')
+parser.add_argument('-bs', '--batch_size', default=2, type=int, help='Batch size of the training and validation processes')
+parser.add_argument('-nw', '--num_workers', default=4, type=int, help='Number of workers for the dataloaders')
+parser.add_argument('-lr', '--learning_rate', default=1e-4, type=float, help='Learning rate for training the model')
+parser.add_argument('-wd', '--weight_decay', type=float, default=0.01, help='Weight decay (i.e. regularization) value in AdamW')
+parser.add_argument('-pat', '--patience', default=250, type=int, help='number of validation steps (val_every_n_iters) to wait before early stopping')
+parser.add_argument('--T_0', default=100, type=int, help='number of steps in each cosine cycle')
+parser.add_argument('-epb', '--enable_progress_bar', default=False, type=bool, help='by default is disabled since it doesnt work in colab')
```
Comment (uzaymacar) [inline modeling/main.py:60]: I would set this to `True` by default. Otherwise, a newcomer to the script might wonder why no progress bar is showing up.

File: modeling/datasets_sc.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+
+                # Convert to Numpy
+                sag_img, gt = sag_img.get_fdata(), gt.get_fdata()
+
+                # Apply center-cropping
+                center_crop = CenterCrop(size=self.center_crop_size)
+                sag_img = center_crop(sample=sag_img, metadata={'crop_params': {}})[0]
+                gt = center_crop(sample=gt, metadata={'crop_params': {}})[0]
+
+                if self.center_crop_size == self.subvolume_size:    # Move on to the prediction
```
Comment (uzaymacar) [inline modeling/datasets_sc.py:281]: Inside this `if` statement, all references to `sag_img_subvolume` should be renamed as `sag_img` as `sag_img_subvolume` is only available for the `else` statement.

File: modeling/datasets_sc.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+                    # Check if padding & un-padding removes any lesion GTs; only continue if it does not
+                    if abs(np.sum(gt) - gt_sum_before_crop) > 1e-6:
+                        # NOTE: Apparently np.sum() can have epsilon differences even with same values!
+                        raise ValueError('Padding & un-padding cropped out lesions! Check your center-crop parameters!')
+            
+                # Calculate the test metrics on the "soft" prediction only
+                for metric_fn in self.metric_fns:
+                    res = metric_fn(pred, gt)
+                    dict_key = metric_fn.__name__
+                    self.test_metrics[dict_key].append(res)
```
Comment (uzaymacar) [inline modeling/datasets_sc.py:341]: Some scores were NaN (we should also investigate why) so we can remove them in performance computation with:
```suggestion
                    if dict_key == 'dice_score' and np.isnan(res):
                        continue

                    self.test_metrics[dict_key].append(res)
```

File: modeling/datasets_sc.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+
+import torch
+from torch.utils.data import Dataset, random_split
+
+from ivadomed.transforms import CenterCrop, RandomAffine, ElasticTransform, \
+     NormalizeInstance, RandomReverse, RandomGamma, RandomBiasField, RandomBlur
+from ivadomed.metrics import dice_score, hausdorff_score, intersection_over_union, \
+    precision_score, recall_score, specificity_score, numeric_score
+from utils import volume2subvolumes, subvolumes2volume
+
```
Comment (uzaymacar) [inline modeling/datasets_sc.py:19]: It is important to make sure that the loss is computed the same way for training, validation, and testing. You'll notice that whereas in training and validation we use `DiceLoss` from `ivadomed.losses` as seen in `main.py`, we use `dice_score` from `ivadomed.metrics` for testing. Apparently, there is a difference between the two implementations in ivadomed, namely the addition of `smooth` in `DiceLoss` to handle NaN cases. Therefore, here I would change `dice_score` in the following way:
```suggestion
from ivadomed.losses import DiceLoss
dice_score = DiceLoss(smooth=1.0)
dice_score.__setattr__('__name__', 'dice_score')
```

File: modeling/main.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+import torch.nn as nn
+import torch.optim as optim
+import pytorch_lightning as pl
+from torch.nn import functional as F
+from torch.utils.data import DataLoader, dataloader, random_split
+import torchvision.transforms as transforms
+
+# dataloaders and segmentation models
+from datasets_sc import SCIZurichDataset
+from unet3d import ModifiedUNet3D
```
Comment (uzaymacar) [inline modeling/main.py:22]: I didn't try this yet, but it might be better to import directly from `ivadomed` with 

```
from ivadomed.models import ModifiedUNet3D
```

here to increase consistency between ivadomed and this custom script.

**Review by uzaymacar — CHANGES_REQUESTED** [review]
I have added some minor comments. Now about the performances: with training Dice scores of ~`0.6`, I was only able to get test Dice scores of ~`0.13`. Therefore, this looks like another case of severe overfitting. I wasn't able to find out any error in the data loading process.

A thing I noticed is that even in the `self.subjects` subset (which includes training + validation subjects), there are many predictions with `0.0` Dice scores and the overall Dice score in this set is ~`0.4`. This is surprising as PyTorch Lightning (PL) progress bar shows about ~`0.6` training loss as mentioned above.

Next steps are:
* Making sure the way we compute scores across PL in `main.py` and `test()` function in `datasets_sc.py` are consistent.
* Brainstorm on how to reduce overfitting (e.g., reducing data augmentations)
* Making sure parameters from the ivadomed config file mentioned in this PR are translated correctly to this custom script

File: modeling/datasets_sc.py (hunk 2 above)
Comment (naga-karthik) [reply modeling/datasets_sc.py:281]: Yeah, this is a good point, I was too lazy to modify everything, so I left it as it was. Will update them now!

File: modeling/datasets_sc.py (hunk 4 above)
Comment (naga-karthik) [reply modeling/datasets_sc.py:19]: Ah okay, I hadn't noticed that, thanks for pointing it out! Will update

File: modeling/main.py (hunk 5 above)
Comment (naga-karthik) [reply modeling/main.py:22]: It's exactly the same as the one in ivadomed (and ms-challenge)! Remember how we were thinking of moving away from having a huge models.py file for ivadomed? I wanted to realize that in-practice and hence thought of having different files for different models. But it is the same, shouldn't make any difference. I can add a comment as well saying that it is the same

**Comment by naga-karthik** [comment]
We switched to nnUNet for the sci-model as it is considered the state-of-the-art, hence this PR with the custom model will not be worked on!
